Encode Authentication token

pacejackson opened the pull request:
Thrift expects a header to be a str in Python 2 (str or bytes works in Python 3) so we `encode` the authentication token in the `AuthenticationContext` constructor so `AuthenticationContext._token` is always `bytes` in Python 3 or `str` in Python 2.  

If we do not do this, the value may be `unicode` in Python 2.  `unicode` works fine as long as the length of the string is under 128 characters but once you go over that, you start to run into errors when Thrift tries to construct the header.

pacejackson commented:
Here is what happens before this change:

```python
token = u"x" * 128
auth = AuthenticationContext(token=token)
c = EdgeRequestContext.create(authentication_context=auth)
c.attach_context(context)
thing_service.get_trending_subreddits()

---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-10-ebeea19a7a37> in <module>()
     28 fact = ThriftContextFactory(pool, ThingService.Client)
     29 client = fact.make_object_for_context('test', context)
---> 30 client.get_trending_subreddits()

/usr/lib/python2.7/dist-packages/baseplate/context/thrift.pyc in _call_thrift_method(self, name, *args, **kwargs)
    120                         client = self.client_cls(prot)
    121                         method = getattr(client, name)
--> 122                         return method(*args, **kwargs)
    123             except TTransportException as exc:
    124                 last_error = exc

/home/vagrant/graphql/graphql_api/thing_thrift/ThingService.pyc in get_trending_subreddits(self, token)
   4780      - token
   4781     """
-> 4782     self.send_get_trending_subreddits(token)
   4783     return self.recv_get_trending_subreddits()
   4784 

/home/vagrant/graphql/graphql_api/thing_thrift/ThingService.pyc in send_get_trending_subreddits(self, token)
   4789     args.write(self._oprot)
   4790     self._oprot.writeMessageEnd()
-> 4791     self._oprot.trans.flush()
   4792 
   4793   def recv_get_trending_subreddits(self, ):

/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.pyc in flush(self)
    372 
    373     def flush(self):
--> 374         self.flushImpl(False)
    375 
    376     def onewayFlush(self):

/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.pyc in flushImpl(self, oneway)
    471         buf = StringIO()
    472         if self.__client_type == self.HEADERS_CLIENT_TYPE:
--> 473             self._flushHeaderMessage(buf, wout, wsz)
    474         elif self.__client_type == self.FRAMED_DEPRECATED:
    475             buf.write(pack(b"!I", wsz))

/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.pyc in _flushHeaderMessage(self, buf, wout, wsz)
    410         THeaderTransport._flush_info_headers(info_data,
    411                                              self.__write_headers,
--> 412                                              self.INFO_KEYVALUE)
    413 
    414         header_data = StringIO()

/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.pyc in _flush_info_headers(info_data, write_headers, type)
    531             for str_key, str_value in write_headers_iter:
    532                 info_data.write(THeaderTransport._serialize_string(str_key))
--> 533                 info_data.write(THeaderTransport._serialize_string(str_value))
    534             write_headers.clear()
    535 

/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.pyc in _serialize_string(str_)
    518         if sys.version_info[0] >= 3 and not isinstance(str_, bytes):
    519             str_ = str_.encode()
--> 520         return getVarint(len(str_)) + str_
    521 
    522     @staticmethod

UnicodeDecodeError: 'ascii' codec can't decode byte 0x80 in position 0: ordinal not in range(128)
```

ckwang8128 commented:
:nail_care: is there a test already to verify the Thrift encoding/decoding pipeline works with this? if not, could one be added?

pacejackson commented:
:haircut_man: -ish

I'm still thinking through the requested test, but when I was thinking about this over the weekend, I was thinking that a better approach is to `encode` the token in the `AuthenticationContext` constructor so we know that it is always the same type.  What do you think of that approach?

pacejackson commented:
I added an integration test for this, here is the output of the tests I added when the changes to `AuthenticationContext` are commented out:

Python 2:
```
======================================================================
ERROR: test_client_proxy_flow (tests.integration.thrift_tests.ThriftTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/baseplate/tests/integration/thrift_tests.py", line 291, in test_client_proxy_flow
    client.example_simple()
  File "/home/vagrant/baseplate/baseplate/context/thrift.py", line 122, in _call_thrift_method
    return method(*args, **kwargs)
  File "/home/vagrant/baseplate/tests/integration/test_thrift/TestService.py", line 448, in example_simple
    self.send_example_simple()
  File "/home/vagrant/baseplate/tests/integration/test_thrift/TestService.py", line 456, in send_example_simple
    self._oprot.trans.flush()
  File "/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.py", line 374, in flush
    self.flushImpl(False)
  File "/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.py", line 473, in flushImpl
    self._flushHeaderMessage(buf, wout, wsz)
  File "/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.py", line 412, in _flushHeaderMessage
    self.INFO_KEYVALUE)
  File "/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.py", line 533, in _flush_info_headers
    info_data.write(THeaderTransport._serialize_string(str_value))
  File "/usr/lib/python2.7/dist-packages/thrift/transport/THeaderTransport.py", line 520, in _serialize_string
    return getVarint(len(str_)) + str_
UnicodeDecodeError: 'ascii' codec can't decode byte 0x85 in position 0: ordinal not in range(128)

======================================================================
FAIL: test_python_2_ensure_token_is_str (tests.unit.core_tests.AuthenticationContextTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/baseplate/tests/unit/core_tests.py", line 316, in test_python_2_ensure_token_is_str
    self.assertIs(type(auth_context._token), str)
AssertionError: <type 'unicode'> is not <type 'str'>
```

Python 3:
```
======================================================================
FAIL: test_python_3_ensure_token_is_bytes (tests.unit.core_tests.AuthenticationContextTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/baseplate/tests/unit/core_tests.py", line 306, in test_python_3_ensure_token_is_bytes
    self.assertEqual(auth_context._token, self.VALID_TOKEN)
AssertionError: 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ[217 chars]-38I' != b'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.ey[218 chars]-38I'
```

pacejackson commented:
:haircut_man:

jdost commented on baseplate/core.py:
```diff
@@ -168,6 +169,16 @@ class AuthenticationContext(object):
+        # of the value being set as a header is over 128 characters and the
+        # value is of type unicode.  Rather than coverting before passing as a
+        # header, we encode in the constructor so we know that self._token is
+        # always the same type.
```
:trophy:
